src/nnet3/nnet-training.cc

```diff
@@ -118,6 +104,64 @@ void NnetTrainer::ProcessOutputs(const NnetExample &eg,
   }
 }
 
+void NnetTrainer::UpdateParamsWithMaxChanges() {
+  KALDI_ASSERT(delta_nnet_ != NULL);
+  // computes scaling factors for per-component max-changes
+  const int32 num_ucs = NumUpdatableComponents(*delta_nnet_);
+  Vector<BaseFloat> scale_factors = Vector<BaseFloat>(num_ucs);
+  BaseFloat param_delta = 0.0;
+  int32 i = 0;
+  for (int32 c = 0; c < delta_nnet_->NumComponents(); c++) {
+    Component *comp = delta_nnet_->GetComponent(c);
+    if (comp->Properties() & kUpdatableComponent) {
+      UpdatableComponent *uc = dynamic_cast<UpdatableComponent*>(comp);
+      if (uc == NULL)
+        KALDI_ERR << "Updatable component does not inherit from class "
+                  << "UpdatableComponent; change this code.";
+      BaseFloat max_param_change_per_comp = uc->MaxChange();
+      KALDI_ASSERT(max_param_change_per_comp >= 0.0);
+      BaseFloat dot_prod = uc->DotProduct(*uc);
+      if (max_param_change_per_comp != 0.0 &&
+          std::sqrt(dot_prod) > max_param_change_per_comp) {
+        scale_factors(i) = max_param_change_per_comp / std::sqrt(dot_prod);
+        num_max_changes_per_component_applied_[i]++;
+        KALDI_VLOG(2) << "Parameters in " << delta_nnet_->GetComponentName(c)
+                      << " change too big: " << std::sqrt(dot_prod) << " > "
+                      << "--max-change=" << max_param_change_per_comp
+                      << ", scaling by " << scale_factors(i);
+      } else {
+        scale_factors(i) = 1.0;
+      }
+      param_delta += std::pow(scale_factors(i), 2.0) * dot_prod;
+      i++;
+    }
+  }
+  KALDI_ASSERT(i == scale_factors.Dim());
+  param_delta = std::sqrt(param_delta);
+  // computes the scale for global max-change (with momentum)
+  BaseFloat scale = (1.0 - config_.momentum);
+  if (config_.max_param_change != 0.0) {
+    param_delta *= scale;
+    if (param_delta > config_.max_param_change) {
+      if (param_delta - param_delta != 0.0) {
+        KALDI_WARN << "Infinite parameter change, will not apply.";
+        SetZero(false, delta_nnet_);
+      } else {
+        scale *= config_.max_param_change / param_delta;
+        num_max_changes_global_applied_++;
+        KALDI_LOG << "Parameters change too big: " << param_delta << " > "
```

Comment:
I think you should rework how the log messages are printed.  Here you print out a log message per minibatch if you apply the global max change, but not if you apply the per-component max change.  If the quota of verboseness is one log message of reasonable length per minibatch, then you could do a better job of compressing more information in that.  How about you print out one message if *any* max-change was applied, and find a way to neatly summarize how many components had a per-component max-change limit applied (and maybe print the component-name of the one that had the smallest scale applied, together with its max-change value), and also say what the global max-change limit was, if any, and what the global max-change was.